### ISSUE TITLE
Strange behaviour with dfs_tree and depth_limit

### ISSUE THREAD

**Issue by WiseNat** [networkx/networkx#6517]
![image](https://user-images.githubusercontent.com/56839545/226110730-bd02a05c-1a76-4803-8735-64e6da9075f9.png)

```py
import networkx as nx

edges = {
    "A": ["B", "C"],
    "D": ["C"],
    "C": ["D", "A", "B"],
    "B": ["C", "A", "E"],
    "E": ["B"]
}


G = nx.Graph(edges)

source = "A"
depth_limit = 2

dfs_result = nx.dfs_tree(G, source=source, depth_limit=depth_limit)
bfs_result = nx.bfs_tree(G, source=source, depth_limit=depth_limit)

print(f"DFS: {dfs_result.nodes}")
print(f"BFS: {bfs_result.nodes}")

>>> DFS: ['A', 'B', 'C', 'E']
>>> BFS: ['A', 'B', 'C', 'E', 'D']
```

`dfs_tree` doesn't return node D for whatever reason when edge BC exists despite returning an node E which is equidistant. If that edge is removed then all nodes are returned.

This issue doesn't occur with `bfs_tree` in any situation.

Not sure if this is intended behaviour or not but it just seems odd to me.

**Comment by dschult** [comment]
This might be related to #6479 if it is related to the depth_limit not yielding nodes that it hasn't completely explored. It turns out that defining what the "depth" is can be done in multiple ways when there is a depth_limit. At depth 5 should you yield all nodes that you reach (preorder) or all the nodes you would have explored if the depth_limit was not in place (middle level order -- I don't know a name for it) or all the nodes that you successfully explore (postorder).

If this Issue is the same problem as #6479 then we need to change the behavior of the depth_limited version of the dfs functions.

**Comment by brocla** [comment]
Interesting. I stepped through the code. Found this.

**D** isn't found because **D** is at depth 3 when searched in this  order **A** -> **B** -> **C** -> **D**, which is how `dfs_tree()` starts its traversal.

I expected that it would find **D** later when it traversed along  **A** ->  **C** -> **D** but this doesn't happen because **C** has already been marked as 'visited' from the first traversal. The algorithm ignores the path thru **C** and everything beyond it.

Hmmm...

**Comment by brocla** [comment]
`dfs_tree()` calls `dfs_edges()` to do the work of traversing. The issue is actually in `dfs_edges()`.

Curiously, the documentation for `dfs_edges()` says,

>  may not generate all edges in G.

Which, in this case, turns out to be true. I do not know why that would be desirable.

**Comment by dschult** [comment]
It is definitely true that the order of the nodes at each level will change the depth of the nodes in a DFS. So, the depth-limit can affect quite a lot. 

The reason for not generating all the edges is that the DFS is often used for things that don't want all the edges. For example, DFS edges will form a tree, leaving out the edges that form cycles.

**Comment by brocla** [comment]
Thanks for the explanation!

Kevin

**Comment by Mamuschka** [comment]
> This might be related to #6479 if it is related to the depth_limit not yielding nodes that it hasn't completely explored. It turns out that defining what the "depth" is can be done in multiple ways when there is a depth_limit. At depth 5 should you yield all nodes that you reach (preorder) or all the nodes you would have explored if the depth_limit was not in place (middle level order -- I don't know a name for it) or all the nodes that you successfully explore (postorder).
> 
> If this Issue is the same problem as #6479 then we need to change the behavior of the depth_limited version of the dfs functions.

nah, this is the only way a dfs-search with deph limit can work.

![image](https://user-images.githubusercontent.com/20465511/226826416-491d7ffb-e208-4b38-b9cc-c6656a29aff5.png)

The yellow and green edges are the dfs-Tree with source A. With deph-limit the Tree have to loose the green edge. There is no other way you can travelise deph-first and get all nodes and have a Tree with deph 2.

This was the reason, I needed the bfs-nodes. Than only in the bfs with deph-limit 2 you reach all nodes that are 2 edges away from the source.
